### DISCUSSION TITLE
Strongly typed ad-hoc shapes

### DISCUSSION THREAD

**Discussion by rjpowers10** [OrchardCMS/OrchardCore#10099] [Q&A] [answered]
The model type for an ad-hoc shape is dynamic. This is true in both Orchard 1 and Orchard Core. However, in O1 we were able to work around it doing something like this:

```
@Display(New.SampleShape(SampleModel: myModel)
```

```
@{
    var myModel = Model.SampleModel;
    var localHtmlHelper = Html.HtmlHelperFor(myModel);   // HtmlHelperFor is an O1 extension method
}


<form>
    @* Use localHtmlHelper instead of global HtmlHelper *@
</form>
```

Not only does the HtmlHelperFor extension method not exist in Orchard Core, but ASP.NET has moved towards using tag helpers instead of HTML helpers. What I want is to be able to do something like this

```
<input asp-for="What goes here?"></input>
```

I want to do this because I have a form element and I want to use asp-for to set the names of the inputs.

I suspect there isn't a good answer here other than setting the names of the inputs manually, but I figured I would ask to see if anyone has any ideas.

Thanks!

**Comment by ns8482e** [comment]
For typed shapes - Drivers are the way to go, You can define fully typed view model from Driver, And use them in View using asp tag helpers.  you can use POCO or you can also inherit your shape models from `ShapeViewModel`

**Comment by Piedone** [comment, marked as answer]
Just a nitpick: What you're asking about is not a "strongly" but a "statically" typed model.

This is not too easy but you can create statically typed shapes like this: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/issue/OSOE-41/Views/DisplayManagement/AdHocShape.cshtml#L19 (This is what drivers also do.)

**Reply by rjpowers10** [reply]
Thanks for the reply! I'm marking this as the answer although @mroskamp's answer is equally valid :)

**Comment by mroskamp** [comment]
You could achieve a similar result by deriving your statically typed model from `ShapeViewModel`. There are a few examples of this in the Orchard Core code, like `SearchFormViewModel`. `ShapeViewModel` is an `IShape` and takes a shape type at construction time, so you can construct your derived shape, specifying your desired shape type, and pass that directly into `DisplayAsync` in your view template.

**Reply by Piedone** [reply]
Nice! I've added a tutorial about this too: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/issue/OSOE-41/Views/DisplayManagement/AdHocShape.cshtml#L36

**Reply by Skrypt** [reply]
ShapeViewModel is the cleanest way. We surely should use this design more often but refactoring everything in OC would take a lot of time to do so. But, definitely cleaner using a "strongly" typed ViewModel than a "statically" one.
